nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+		}
+	}
+	e.Status = stResolved
+	e.Flags = flags
+	e.putTime = now
+	e.validUntil = now.Add(time.Second * time.Duration(minTtl))
+	e.reply = *reply
+
+	if shouldNotify {
+		close(e.waitChan) // notify all the waiters by closing the channel
+	}
+}
+
+// wait until a valid reply is set in the cache
+func (e *cacheEntry) waitReply(request *dns.Msg, timeout time.Duration, now time.Time) (*dns.Msg, error) {
+	if e.Status == stResolved {
+		return e.getReply(request, now)
+	}
+
+	if timeout > 0 {
+		select {
+		case <-e.waitChan:
+			return e.getReply(request, now)
+		case <-time.After(time.Second * timeout):
+			return nil, errTimeout
+		}
+	}
+
+	return nil, errCouldNotResolve
+}
+
+func (e *cacheEntry) close() {
+	if e.Status == stPending {
+		close(e.waitChan)
+	}
+}
+
+// entriesSlice is used for sorting entries
+type cacheEntriesSlice []*cacheEntry
+
+func (p cacheEntriesSlice) Len() int           { return len(p) }
+func (p cacheEntriesSlice) Less(i, j int) bool { return p[i].validUntil.Before(p[j].validUntil) }
+func (p cacheEntriesSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+type cacheKey struct {
+	question dns.Question
+	protocol dnsProtocol
+}
+
+type entries map[cacheKey]*cacheEntry
+
+// Cache is a thread-safe fixed capacity LRU cache.
+type Cache struct {
+	Capacity int
+
+	entries entries
+	lock    sync.RWMutex
+}
+
+// NewCache creates a cache of the given capacity
+func NewCache(capacity int) (*Cache, error) {
+	if capacity <= 0 {
+		return nil, errInvalidCapacity
+	}
+	c := &Cache{
+		Capacity: capacity,
+		entries:  make(entries, capacity),
+	}
+	return c, nil
+}
+
+// Clear removes all the entries in the cache
+func (c *Cache) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(entries, c.Capacity)
+}
+
+// Purge removes the old elements in the cache
+func (c *Cache) Purge(now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.removeOldest(c.Capacity, now)
+}
+
+// Add adds a reply to the cache.
+func (c *Cache) Put(request *dns.Msg, reply *dns.Msg, proto dnsProtocol, flags uint8, now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		ent.setReply(reply, flags, now)
+	} else {
+		// If we will add a new item and the capacity has been exceeded, make some room...
+		if len(c.entries) >= c.Capacity {
+			c.removeOldest(1, now)
+		}
+		c.entries[key] = newCacheEntry(&question, reply, stResolved, now)
+	}
+}
+
+// Look up for a question's reply from the cache.
+// If no reply is stored in the cache, it returns a `nil` reply and no error. The caller can then `Wait()`
+// for another goroutine `Put`ing a reply in the cache.
+func (c *Cache) Get(request *dns.Msg, proto dnsProtocol, now time.Time) (reply *dns.Msg, err error) {
```

Comment:
> In conclusion, I would consider your algorithm an optimization more than the base case.

I'm not as concerned with optimisation as I am with simplicity, so I would prefer to say that it's removing complication (e.g., protocol arguments to procedures).

If you include changing the upstream client buffer size back to 512B, then I can see that your scheme works. (By the way you are entirely right that we don't truncate responses -- and the dns library doesn't either so far as I can tell).

Let me try to remove some complication, though: wrt to UDP and TCP, all you really care about is whether you might need to send a truncated result -- with TCP you never need to, and with UDP you need to if the upstream server needed to (assuming both downstream and upstream clients have the same size buffer, i.e., 512B). That's why two different answers might be cached, and the protocol is threaded through methods, to the cache.

Instead of storing two answers, you could look at whether the answer needs to be truncated *when you go to write it out*. You seem to think this is a major bit of computation, but it's really not: you just need to look at the length. And to truncate it, you can just set the bit and remove all RRs (or just don't copy them over from the cached result!). None of this needs to be done inside the cache code.

(It's not a problem to send an empty answer in a truncated response: that's what djbdns does, and miekg/dns doesn't even bother parsing the RRs of packets with the truncate bit set.)